### PR TITLE
add rater for generated answer

### PR DESCRIPTION
**Pull request by Panzy-18**
+ Write a new class RaterForGeneratedAnswerConfig in the config file.
+ Write a new notebook to show how to use and give 3 examples.
+ The prompt of this config may need elaborately designed.

### CODE REVIEW

File: uniflow/flow/config.py [hunk 1]
```diff
@@ -186,6 +186,93 @@ def check_labels_in_label2score(self) -> bool:
...
+    guided_prompt_template: GuidedPrompt = GuidedPrompt(
+        instruction="""Rate the generated answer compared to the grounding answer to the question. Accept means the generated answer is better than the grounding answer and reject means worse.
+        Follow the format of the examples below to include context, question, grounding answer, generated answer and label in the response.
+        The response should not include examples in the prompt.""",
+        examples=[
+            Context(
+                context="Basic operating system features were developed in the 1950s, such as resident monitor functions that could automatically run different programs in succession to speed up processing.",
+                question="When were basic operating system features developed?",
+                grounding_answer="Basic operating system features were developed in the 1980s.",
+                generated_anser="Basic operating system features were developed in the 1950s",
```
Comment (goldmermaid) [inline uniflow/flow/config.py:215]: The `generated_answer` and `grounding_answer` are supposed to be different, so the LLM can compare the difference.

Comment (goldmermaid) [reply uniflow/flow/config.py:215]: `generated_anser` typo

File: uniflow/flow/config.py [hunk 2]
```diff
@@ -186,6 +186,93 @@ def check_labels_in_label2score(self) -> bool:
...
+        """
+        Check if every label in the guided_prompt_template's examples is a key in label2score.
+
+        Returns:
+            bool: True if all labels are in label2score, False otherwise.
+        """
+        for example in self.guided_prompt_template.examples:
+            if example.label not in self.label2score:
+                return False
+        return True
```
Comment (goldmermaid) [inline uniflow/flow/config.py:273]: nit: you can refactor `__post_init__` and `check_labels_in_label2score` to the base RaterConfig for proper OOP design without duplicate code across multiple child class.

File: uniflow/flow/config.py (hunk 1 above)
Comment (Panzy-18) [reply uniflow/flow/config.py:215]: Actually the grounding_answer is "1980s" and generated_answer is "1950s", but I do think this example may not be too appropriate.

File: example/rater/classification.ipynb [hunk 3]
```diff
@@ -133,6 +152,13 @@
     "client = RaterClient(config)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then we can run this client. Client will sample output from LLM 3 times to generate final result."
```
Comment (goldmermaid) [inline example/rater/classification.ipynb:159]: Let' add a header here named "Run the client". Also, we can add more details such as "Then we can run the client. For each item in the `raw_input`, the Client will generate an explanation and a final label `Yes` or `No`. The label is decided by taking the majority votes from sampling the LLM output 3 times, which improved stability compared with outputting 1 time.

File: example/rater/classification.ipynb [hunk 4]
```diff
@@ -1,8 +1,31 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Use AutoRater to classify answer label from a Jupyter Notebook\n",
```
Comment (goldmermaid) [inline example/rater/classification.ipynb:7]: Can you change the title to `Use AutoRater to Assess Question Answer Accuracy`

File: example/rater/generated_answer.ipynb [hunk 5]
```diff
@@ -0,0 +1,355 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Use AutoRater to compare a generated answer to grounding answer from a Jupyter Notebook\n",
```
Comment (goldmermaid) [inline example/rater/generated_answer.ipynb:7]: Can you change the title to "Use `AutoRater` to Compare Answers to a Given Question"

File: example/rater/generated_answer.ipynb [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    "    }\n",
+    ")\n",
+    "client = RaterClient(config)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then we can run this client. Client will sample output from LLM 5 times to generate final result."
```
Comment (goldmermaid) [inline example/rater/generated_answer.ipynb:166]: Let's add a header here named "Run the client". 

Also, we can add more details such as "Then we can run the client. For each item in the `raw_input`, the Client will generate an explanation and a final label [`Strong accept`, `Accept`, `Equivalent`, `Reject`,`Strong reject`] . The label is decided by taking the majority votes from sampling the LLM output 5 times, which improved stability compared with outputting 1 time.

File: example/rater/generated_answer.ipynb [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "data 0 has majority vote accept and average score 1.0\n",
+      "data 1 has majority vote reject and average score -1.2\n",
+      "data 2 has majority vote equivalent and average score 0.0\n"
```
Comment (goldmermaid) [inline example/rater/generated_answer.ipynb:322]: Is there a way to highlight `accept`, `reject` and `equivalent` and their scores? e.g. 

"data 1 has majority vote `REJECT` and average score `-1.2` over five-time LLMs' simulations.

File: example/rater/classification.ipynb [hunk 8]
```diff
@@ -1,8 +1,31 @@
...
+    "\n",
+    "In this example, we will show you how to use autorater to classify answer label from a given jupyter notebook.\n",
```
Comment (goldmermaid) [inline example/rater/classification.ipynb:9]: "In this example, we will show you how to use `AutoRater` to verify the correctness of an answer to a given question and context pairs."

File: example/rater/json_formatted_classification.ipynb [hunk 9]
```diff
@@ -1,8 +1,31 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Use AutoRater to classify answer label from a Jupyter Notebook\n",
+    "\n",
+    "In this example, we will show you how to use autorater to classify answer label from a given jupyter notebook.\n",
```
Comment (goldmermaid) [inline example/rater/json_formatted_classification.ipynb:9]: Can you adjust this notebook similar to what I commented in the previous ones?

**Comment by Panzy-18** [comment]
Also I find a bug that if I change` response_format={"type": "text"}` to `response_format={"type": "json_object"}` in `generated_answer.ipynb`, GPT3.5 seems to ignore my prompt to not include any example in prompt, thus can not parse a right result.

**Comment by Panzy-18** [comment]
old prompt:

|             | GPT3.5            | GPT4              |
| ----------- | ----------------- | ----------------- |
| text        | expected response | expected response |
| json_object | wrong response    | expected response |

I write a new prompt for this task. This prompt is longer (consume more tokens) but more informative.

|             | GPT3.5                                 | GPT4              |
| ----------- | -------------------------------------- | ----------------- |
| text        | expected response                      | expected response |
| json_object | right format, but low self-consistency | expected response |

Also I obeserved that GPT-4 have a higher confidence and self-consisitency for each input than GPT-3.5

File: example/rater/json_formatted_classification.ipynb
Comment (goldmermaid) [inline example/rater/json_formatted_classification.ipynb:1]: it seems that the json formatted notebook is removed in your PR?

Comment (Panzy-18) [reply example/rater/json_formatted_classification.ipynb:1]: Yes, I merged them in one notebook.

Comment (goldmermaid) [reply example/rater/json_formatted_classification.ipynb:1]: Cool! Can you resolve the conflicts below?  in the `uniflow/flow/config.py` file

File: uniflow/flow/config.py [hunk 10]
```diff
@@ -137,24 +137,38 @@ class RaterConfig:
...
-        for example in self.guided_prompt_template.examples:
-            if example.label.lower() not in [k.lower() for k in self.label2score]:
-                raise ValueError(
-                    "Inconsistent labels found in guided_prompt_template examples, "
-                    f"example label {example.label} not in label2score has keys {list(self.label2score.keys())}",
-                )
-
-    def check_labels_in_label2score(self) -> bool:
+        incompatible_labels = self.check_labels()
+        unexprected_labels = incompatible_labels["unexpected_labels"]
```
Comment (goldmermaid) [inline uniflow/flow/config.py:141]: qq: is `unexprected_labels` a spelling error?

**Review by CambioML — APPROVED** [review]
LGTM.

File: uniflow/flow/config.py (hunk 10 above)
Comment (CambioML) [reply uniflow/flow/config.py]: addressed directly through github.

File: example/rater/json_formatted_classification.ipynb
Comment (CambioML) [reply example/rater/json_formatted_classification.ipynb:1]: addressed directly through github

**Pull request merged**